### PR TITLE
fix(metadata-store): #916 Ensure drop transaction gets committed

### PR DESCRIPTION
**Pull request by ncloudioj**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-stream/917)

### CODE REVIEW

**Comment by ncloudioj** [comment]
@oyiptong r?

@sarracini and @k88hudson have found the `asycnDrop` was a bit annoying if it couldn't guarantee the drop transaction gets committed. So I made a workaround to make it easier to use.

File: lib/MetadataStore.js [hunk 1]
```diff
@@ -280,6 +281,38 @@ MetadataStore.prototype = {
   }),
 
   /**
+   * This function could be used to ensure the "asyncDrop" actually commit
+   * the transaction. It appears that the transaction might be still in
+   * the uncommitted state despite its promise is resolved in Sqlite.jsm.
+   * Hence, it has to poll for the table info periodically to comfirm that
+   */
+  _waitForDrop() {
```
Comment (oyiptong) [inline lib/MetadataStore.js]: please rename, it's not waiting for a drop, it's waiting for a reset

**Comment by oyiptong** [comment]
from our conversation IRL: upon uninstall/disable, it is preferable to delete the database rather than reset the database, otherwise we'd leave an empty (but fully created) sqlite database  behind

**Comment by ncloudioj** [comment]
Very good points. Will make those changes accordingly.

**Comment by ncloudioj** [comment]
@sarracini FYI, I've refactored the MetadataStore a bit, specifically,
* rename `asyncDrop` to `asyncReset`, it resets all the tables (essentially, it does a combo - "drop" and then "create tables"
* add `asyncDelete` to nuke the SQLite file, I've also updated the code that handles the "disable/uninstall" events in `main.js`

**Comment by ncloudioj** [comment]
@oyiptong One more round, please? :)

File: lib/MetadataStore.js [hunk 2]
```diff
@@ -280,21 +282,70 @@ MetadataStore.prototype = {
   }),
 
   /**
+   * This function could be used to ensure the "asyncReset" actually commit
+   * the transaction. It appears that the transaction might be still in
+   * the uncommitted state despite its promise is resolved in Sqlite.jsm.
+   * Hence, it has to poll for the table info periodically to comfirm that
+   */
+  _waitForAsyncReset() {
```
Comment (oyiptong) [inline lib/MetadataStore.js]: makes sense to move this back to the `test/lib/utils.js` now right?

**Comment by oyiptong** [comment]
r+ with a nit: please move the `waitForReset` function back to `test/lib/utils.js`

File: lib/MetadataStore.js (hunk 2 above)
Comment (ncloudioj) [inline lib/MetadataStore.js]: Sure, will do.

**Pull request merged**